Refactor xbgpu to support multiple output streams

amishatishpatel opened the pull request:
Much like #537, the result of this refactoring (and added unit testing endeavours) means the plumbing is in and (mostly?) verified - at least for the case of running multiple XPipelines.

The main change is around moving the gpu-processing and sender loops into an base Pipeline class, to be implemented by subclasses. As mentioned above, I've created an XPipeline (as per NGC-996) which contains most of the existing xbgpu codebase - just a bit reorganised and tweaked where necessary.
* A single `XBEngine` manages multiple `Pipeline`s,
* Each Pipeline corresponds to an `Output`,
* Each Pipeline also has its own processing and download command queues.

Currently only the `XPipeline` is supported (its also the only one that exists).

As mentioned in the last commit before raising this PR, `test_engine` requires a bit of cleaning up, but the structure is there. I might take further inspiration from @bmerry and opt to update relevant documentation at the end of this adventure.

As always, please and thank you.

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [ ] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [ ] (N/A) If modules are added/removed: use sphinx-apidoc to update files in `doc/`
- [ ] Ensure copyright notices are present and up-to-date
- [ ] (N/A) If qualification tests are changed: attach a sample qualification report
- [ ] If design has changed: ensure documentation is up to date
- [ ] If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

Contributes to NGC-929 and Closes NGC-996.

james-smith-za commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -318,14 +749,14 @@ def __init__(
         n_rx_items = 3  # Too high means too much GPU memory gets allocated
-        n_tx_items = 2
 
         # Multiply this _step by 2 to account for dropping half of the
         # spectrum due to symmetric properties of the Fourier Transform. While
         # we can workout the timestamp_step from other parameters that
         # configure the receiver, we pass it as a seperate argument to the
         # reciever for cases where the n_channels_per_stream changes across
         # streams (likely for non-power-of-two array sizes).
+        # NOTE: Should be the same for both X- and B-engines, right?
```
The thought occurs to me to wonder whether this comment actually relates to the code anymore. It may be worth digging back into the history. There's no 2 here, was there at one point?

bmerry replied:
And yes, n_channels_per_stream will be the same for all streams because it's determined by the input that the engine subscribes to (and non-power-of-two array sizes have nothing to do with it).

james-smith-za commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -396,96 +817,34 @@ def __init__(
         )
+        for _ in range(n_rx_items):
+            # TODO: Abstract dtype, as per NGC-1000
+            buffer_device = accel.DeviceArray(context, rx_data_shape, dtype=np.int8)
```
You've changed the way that the `buffer_device` is assigned here. Why?

Do you want to avoid reaching into one of the Pipeline objects and using the slots' `allocate_host` function?

bmerry replied:
I'm happy with this change. Once we have B pipelines, it will be more difficult to generically access the appropriate slots within the pipelines.

james-smith-za reviewed (changes requested):
I am broadly happy with the approach taken, but I think we need a little bit of tidying up.

There might be one or two points of discussion here, perhaps we can do a zoom at some point to clarify things, but for the time being you should have some things to work on.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -396,96 +817,34 @@ def __init__(
         # * The _tx_item_queue passes items from the _gpu_proc_loop to the
         #   _sender_loop function.
```
There is no tx_item_queue here any more.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -396,96 +817,34 @@ def __init__(
-                buffers=buffers,
-            ),
-            tx_enabled=self._init_tx_enabled,
+        # it back to the corresponding _(rx/tx)_free_item_queue to ensure that
```
No tx_free_item_queue here any more.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
+    async def _add_rx_item(self, item: RxQueueItem) -> None:
+        """Push an :class:`RxQueueItem` to all the pipelines.
+
+        Allow for None to be passed through to indicate shutdown procedure.
```
Is this meant to be a TODO? It doesn't match the type annotation or the body of the function.

amishatishpatel replied:
Ah, legacy from my own reminder (now done with the `shutdown`). Will remove.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -528,254 +920,45 @@ async def _receiver_loop(self) -> None:
-        the network could be pipeline a bit better, but this is not really
-        required in this loop as this whole process occurs at a much slower
-        pace than the rest of the pipeline.
+        The request requires a stream name to enable
```
The generated help for katcp requests doesn't show the function signature, so it's recommended to use a `Parameters` section to formally document the parameter.

amishatishpatel replied:
Rather embarassingly, I didn't know you could `?help <request_name>` like you can with ipython `thing.function?`.

Updating now!

bmerry commented on src/katgpucbf/xbgpu/engine.py:
Again, document the parameter.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
Not quite sure I follow. But rather than exposing the XSend object or its B equivalent directly to the Engine, what about have the base class provide an abstract function to enable or disable transmission?

amishatishpatel replied:
mypy, as always, is upset because it doesn't know about the `send_stream` attribute that we're accessing below.

Your suggestion is a good one indeed, and probably a bit cleaner than reaching straight into a `pipeline` and toggling a value.

bmerry reviewed (changes requested):
I've looked over the src/ tree so far. As James said, the overall structure seems sound, but a few things need fixing and there is a bit of polishing.

I'll come back for the test/ tree tomorrow, so you can leave the ticket assigned to me for now.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -506,46 +523,49 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-        assert prom_diff.get_sample_diff("output_x_bytes_total") == (
-            xbengine.send_stream.heap_payload_size_bytes * n_total_accumulations
+        assert prom_diff.get_sample_diff("output_x_bytes_total", {"stream": "bcp1"}) == (
+            xbengine._pipelines[0].send_stream.heap_payload_size_bytes * n_total_accumulations
```
Could use `_request_pipeline` to do the lookup by name, to ensure that you're querying the right pipeline i.e. the one matching `{"stream": "bcp1"}`.

amishatishpatel replied:
I've changed the way these final checks are done to only loop over XPipelines in `xbengine._pipelines` - I use `pipeline.output.name` when looking up stats.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -23,7 +23,7 @@
 .. todo::
 
-    - The B-Engine logic has not been implemented yet - this needs to be added
+    * The B-Engine logic has not been implemented yet - this needs to be added
```
Apologies, I had actually changed them in an attempt to resolve some doc building errors (which still exist).

Will roll back to `-` convention.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+        * TODO: Figure out whether the `is not None` check on the Chunk is necessary
+        * Apply GPU processing to data in the RxQueueItem
+            * Bind input buffer(s) accordingly
+        * Obtain a free TxQueueItem of the tx_free_item_queue
```
Yeah, I likely meant **off** the queue, but **from** the queue sounds better.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+        The ratio of rx_items to tx_items is not one to one. There are expected
+        to be many more rx_items in for every tx_item out.
+        """
+        rx_item: RxQueueItem | None
```
Trust me, I didn't enjoy adding it either 😅 Something in the pre-commit was complaining it didn't know what `rx_item` is below/in the helper function.

The removal results in "Cannot determine type of `rx_item`"
* Maybe I'll add a `type: ignore` below?

bmerry replied:
While it's not pretty, having mypy know the types of things is definitely preferable to sprinkling on `type: ignore`.

james-smith-za replied:
I'm surprised that mypy can't figure it out because it knows what the type of `_rx_item_queue` is, so it should know what kind of thing `get()` produces.

Would it not be possible to add the type hint where `rx_item` is below? like
```
rx_item: RxQueueItem | None = await self._rx_item_queue.get()
```

I didn't test that at all, this is just how I would have done it. What are the down-sides to what I've suggested compared with what you've got here?

(This is a really minor item. Ignore if you even have the slightest inkling of not wanting to bother.)

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+        # present_ants only takes into account batches that have
+        # been seen. If some batches went missing entirely, the
+        # whole accumulation is bad.
+        if tx_item.batches != self.output.heap_accumulation_threshold:  # type: ignore
```
I added it here because mypy complained
> "Output" has no attribute "heap_accumulation_threshold"

bmerry replied:
That's probably because `self.output` is defined in the base class as `Output`, so it doesn't know that it's actually an XOutput. That problem should vanish once you remove `output` from the base class constructor and handle it in XPipeline.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+                break
+            await rx_item.async_wait_for_events()
+            # TODO: Get rid of this explicit reference to the Chunk
+            assert rx_item.chunk is not None  # mypy doesn't like the fact that the chunk is "optional".
```
Too right, too right. Thanks for the heads-up.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+        logger.debug("gpu_proc_loop completed")
+        self._tx_item_queue.put_nowait(None)
+
+    async def sender_loop(self) -> None:  # noqa: D102
```
It does not, I'd added it in my debugging of why <something> was complaining that I hadn't implemented the @abstractmethod properly. I fixed that but failed to remove my "throw everything at the problem to fix it" - removing now.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
+
+    @staticmethod
+    async def _push_recv_chunk(item: RxQueueItem) -> None:
+        """Return chunk to the stream once `events` have fired."""
```
I'd left it in to allude to the `await` below it. I gather this backtick-manner of mentioning things is more specific to function args.

Will update accordingly (to e.g. "once an item's events have fired").

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
+        if item.refcount == 0:
+            # All Pipelines are done with this item
+            self._active_in_sem.release()
+            # Need to wait for any outstanding events before recycling
```
As usual, fair points all around. I don't like answering a question with another question, _but_ I think it may be an issue because
* Does the item still have an event to a/wait from `rx_item.add_marker(self._proc_command_queue)`?
* It's called/invoked just before the call to `free_rx_item` in the gpu_proc_loop.

bmerry replied:
That captures the GPU processing done on the RxItem (the GPU memory), but you don't need to wait for that to reuse the chunk (CPU memory). Once you've copied the chunk to the GPU, the chunk is ready for reuse.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
+        """Return chunk to the stream once `events` have fired."""
+        await item.async_wait_for_events()
+        item.chunk.recycle()  # type: ignore
+        item.chunk = None
```
Ooff, good catch. I think I was being lazy in giving it the entire RxQueueItem. Perhaps I was also naive in ending `free_rx_item` with a `put_nowait`.
* I'm going to split the difference between this comment and the [earlier one](https://github.com/ska-sa/katgpucbf/pull/600/files#r1246647148) (in free_rx_item) for the next round of reviewing.

amishatishpatel commented on test/xbgpu/test_engine.py:
```diff
@@ -365,31 +384,21 @@ def n_samples_between_spectra(self, n_channels_total: int) -> int:  # noqa: D102
         # Will need to be updated for narrowband
```
Too right, I've created [NGC-1014](https://skaafrica.atlassian.net/browse/NGC-1014).

amishatishpatel commented on test/xbgpu/test_engine.py:
```diff
@@ -404,6 +413,15 @@ async def xbengine(
+        engine_arglist: list[str],
+    ) -> AsyncGenerator[XBEngine, None]:
+        """Create and start an engine based on the fixture values."""
+        arglist = list(engine_arglist)  # Copy, to ensure we don't alter the original
```
> Do you have any reason to think you might be altering the original simply by parsing it?

Honestly? No. Your comment in fgpu/test made sense to me (ethically, morally), so I figured might as well employ the same practice here.

bmerry replied:
In that case I was immediately altering the copy. I don't feel it's necessary here if you're not altering the list.

amishatishpatel commented on test/xbgpu/test_engine.py:
```diff
@@ -506,46 +523,49 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-            else:
-                num_batches_in_current_accumulation = heap_accumulation_threshold
-                if missing_antenna is not None:
+        for i in range(len(mock_send_stream)):
```
Indeed indeed. As mentioned during the consult yesterday, this wasn't quite a complete representation of the required `test_engine` updates.

amishatishpatel commented:
Alrighty, I think I've addressed all the comments under `src/`. Bruce's comments/observations under `test/` are, as always, entirely valid, but will take some time to implement (mainly around testing different `heap_accumulation_threshold`s for each `--corrprod`).

With that in mind, would you both (@james-smith-za and @bmerry ) be happy to review the `src/` updates while I continue working on `test_engine` et al?
* I'm going to request a re-review anyway,
* As usual, ignore until you see fit to interrogate.

james-smith-za commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+        self.correlation.zero_visibilities()
+        return tx_item
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
```
Ok so you haven't changed the `noqa` here but I suspect that's because you moved the docstring to the base class. Which is probably fine, but then this function will get nothing of its own in the module documentation.
That may be OK, I haven't considered the implications too deeply.

bmerry replied:
Sphinx is configured to replicate the base class documentation in the derived class when it renders.

james-smith-za commented on test/xbgpu/test_engine.py:
```diff
@@ -177,6 +178,14 @@ def valid_end_to_end_combination(combo: dict) -> bool:
 
+    @pytest.fixture
+    def mock_send_stream_network(self) -> IPv4Network:
+        """Filter the output queues to just those corresponding to the --corrprods.
```
@amishatishpatel, did you address this comment at all, perhaps during your real-time feedback session?

amishatishpatel replied:
I'd mentioned in my [overall comment](https://github.com/ska-sa/katgpucbf/pull/600#issuecomment-1624877017) before re-requesting a review that I haven't implemented changes/updates to `test_engine.py` just yet (it might take another chunk of time Re: Testing different `heap_accumulation_threshold`s).

The more [top-level aspect](https://github.com/ska-sa/katgpucbf/pull/600#discussion_r1246734554) of this has been addressed, abstracting the common (base?) fixture from fgpu and xbgpu into `test/conftest.py`.

bmerry replied:
I'm not following why it's necessary to override the base implementation. Are there UDP send streams being created that you're trying to filter out? Or is it just future-proofing for when you add beamforming?

amishatishpatel replied:
Promise I haven't been ignoring this comment. Yes, largely for future-proofing reasons. Admittedly, I may have included this pre-emptively.

james-smith-za commented on test/xbgpu/test_engine.py:
```diff
@@ -177,6 +178,14 @@ def valid_end_to_end_combination(combo: dict) -> bool:
+    def mock_send_stream_network(self) -> IPv4Network:
+        """Filter the output queues to just those corresponding to the --corrprods.
+
+        Hardcoded to override the default implementation in conftest.py.
```
ditto here

james-smith-za commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -305,27 +721,25 @@ def __init__(
         # These values are not configurable as they have been acceptable for
         # most tests cases up until now. If the pipeline starts bottlenecking,
         # then maybe look at increasing these values.
+        # TODO: This is the same value as in Pipeline above. Abstract into a single declaration.
```
It would be good to attend to this before you merge, methinks.

james-smith-za reviewed (changes requested):
It's getting close. There are some nit-picks and some outstanding things that you may have missed from previous reviews, which I'd like to be addressed before we merge.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -72,24 +73,29 @@
+    stored separately.
+
+    The RxQueueItem also holds a reference count of the number of pipelines still
+    using this item. This is to ensure the item isn't freed before each data
+    product has had a chance to use the received data.
```
A spot of bike-shedding:
```suggestion
    using this item. This is to ensure the item isn't freed before each
    pipeline has had a chance to use the received data.
```

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -72,24 +73,29 @@
     def reset(self, timestamp: int = 0) -> None:
         """Reset the timestamp, events and chunk."""
         super().reset(timestamp=timestamp)
+        self.chunk: recv.Chunk | None = None
```
`__init__` does call reset, via calling `super().__init__(...)`.

It would probably work to put just the type annotation (without the assignment) in `__init__`, and have just the assignment here.

james-smith-za replied:
I'd feel better about that solution.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -164,14 +167,8 @@ def __init__(self, output: Output, engine: "XBEngine", context: AbstractContext,
-
     def add_rx_item(self, item: RxQueueItem) -> None:
-        """Append a newly-received :class:`RxQueueItem`."""
+        """Append a newly-received :class:`RxQueueItem` to the `rx_item_queue`."""
```
```suggestion
        """Append a newly-received :class:`RxQueueItem` to the :attr:`_rx_item_queue`."""
```
(haven't actually tested that that renders/links correctly)

amishatishpatel replied:
This one is proving a tough nut to crack. It's syntactically similar to e.g. [this reference in fgpu](https://github.com/ska-sa/katgpucbf/blob/main/src/katgpucbf/fgpu/engine.py#L144), but fails to include the actual/needed hypertext anchor to `#katgpucbf.xbgpu.engine.Pipeline._rx_item_queue`
* Perhaps the `_rx_item_queue` needs to be mentioned as a class attribute like `n_rx/tx_items`?

Perhaps++ this is part of the reasoning behind fgpu's `Pipeline.add_in_item` docstring simply saying
> Append a newly-received :class:`~.InItem`.

* I should also add, the mention of [fgpu's _out_queue](https://github.com/ska-sa/katgpucbf/blob/main/src/katgpucbf/fgpu/engine.py#L926) is only ever done in 'plain old comments' - no rST/sphinx directives.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -183,21 +180,23 @@ async def gpu_proc_loop(self) -> None:
-            * Bind input buffer(s) accordingly
-        * Obtain a free TxQueueItem of the tx_free_item_queue
-            * Add event marker to wait for the proc_command_queue
-            * Put the prepared TxQueueItem on the tx_item_queue
```
Need a blank line here - it's not rendering properly. Might also need a blank line before "- Add event marker..."

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -206,17 +205,29 @@ async def sender_loop(self) -> None:
         """
         raise NotImplementedError  # pragma: nocover
 
+    @abstractmethod
+    def capture_start(self) -> None:
+        """Enable the transmission of this data product's stream."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def capture_stop(self) -> None:
+        """Disable the transmission of this data product's stream."""
+        raise NotImplementedError  # pragma: nocover
```
You could probably reduce the amount of boilerplate you deal with by combining these into one function that takes a boolean.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -228,8 +239,7 @@ def __init__(
         context: AbstractContext,
         init_tx_enabled: bool,
```
This parameter seems to be unused now.

amishatishpatel replied:
Is it showing up for you at the end of `self.send_stream = XSend(...)`?
* Arg is `tx_enabled=init_tx_enabled`

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -860,28 +841,14 @@ def free_rx_item(self, item: RxQueueItem) -> None:
-                )
-            )
+            # TODO: Don't recycle the chunk in the reset of the Item itself
+            # Need to separate the concerns
```
Is there still something TODO?

amishatishpatel replied:
As far as I remember, this solution of `recycle()` just before/alongside the `item.reset()` isn't the most desirable?
* It was around the mention of "perhaps the Chunk will need to be reused before the RxQueueItem is done with it"
* Basically, the Chunk doesn't 'belong to' the RxQueueItem, so it should be "allowed to recycle" independently from the resetting of the item.

If that makes sense... Either way, perhaps not a great concern for right now?

bmerry replied:
I just didn't want the call to `recycle` to be *inside* `reset`. I'm happy enough with the way things are.

We are delaying the release of chunks somewhat, so I would suggest running a test correlator at S-band rate before merging just to make sure this isn't going to cause any performance problems.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -860,28 +841,14 @@ def free_rx_item(self, item: RxQueueItem) -> None:
-            )
+            # TODO: Don't recycle the chunk in the reset of the Item itself
+            # Need to separate the concerns
+            item.chunk.recycle()  # type: ignore
```
Rather than suppressing mypy, help it out with something like `assert item.chunk is not None`.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -892,11 +859,11 @@ async def _receiver_loop(self) -> None:
-        4. Place the rx_item on _rx_item_queue so that it can be processed downstream.
+        - Wait for a chunk to be assembled on the receiver.
+        - Get a free rx item off of the _rx_free_item_queue.
+        - Initiate the transfer of the chunk from system memory to the buffer
```
I'm not so sure I like conversion of all the numbered lists to bullet lists. Numbered lists imply that things are happening in a specific order, while this bullet list could be taken as an unordered collection.

This documentation doesn't seem to get rendered (presumably because it is a private member), but I think you need a blank line before the list starts to make it correct.

amishatishpatel replied:
I'm absolutely with you; I was (still am) a big proponent of the numbered list. I changed it because I recall (way back) we had a discussion around numbered-list comments and general discomfort around maintaining/keeping them in parity with what the code is actually doing.

Will roll back and update to add a blank line.

bmerry replied:
Yes, there is a balance. But I think that discussion was more about the numbered comments sprinkled through the code, where adding/removing a step would require renumbering things at distant points, and there were also multiple levels of numbering. With this docstring the list is more self-contained.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -554,11 +547,11 @@ class XBEngine(DeviceServer):
-      3. :func:`_sender_loop` - Transfer correlated data to system RAM and then
-         send it out on the network.
+    - :func:`_receiver_loop` - Receive chunks from network and initiate
+        transfer to GPU.
```
This is rendering funny, maybe due to the indentation.

Also `:func:` should be `:meth:` for methods.

amishatishpatel replied:
Please and thank you for the heads-up. Interestingly, I was able to correctly link gpu-proc and sender loops, but the receiver-loop just goes to bold. Is this the expected behaviour?
![image](https://github.com/ska-sa/katgpucbf/assets/25218138/324537a1-09b2-4933-8a90-bdae55cc3a0e)

bmerry replied:
I think when I last looked, the `_receiver_loop` method wasn't appearing in the built documentation (and hence it can't be linked to), but I didn't investigate why.

bmerry replied:
The mis-indentation is still messing up the rendering.

amishatishpatel commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+
+    def __init__(self, output: Output, engine: "XBEngine", context: AbstractContext, init_tx_enabled: bool) -> None:
+        self.engine = engine
+        self.output = output
```
Indeed, apologies for that. Do you think it's worth making this base Pipeline default to accepting `list[Output]`
* Then, in XPipeline, only take the first element of that list?
* This might add unnecessary complexity/force me to think about number of/relationship between `rx_item_queue`s vs `tx_item_queue`s.

I'm separately trying to figure out the neatest way of letting the base Pipeline (and `add_rx_item`, `shutdown`) know about the `rx_item_queue` - without adding a/nother `type: ignore`.

bmerry replied:
> Do you think it's worth making this base Pipeline default to accepting list[Output]

It depends. Does the base class (or anything that interacts with pipelines without knowing what sort they are) need to access the (list of) outputs? If so, then your approach is what I'd go with. If not, you could just eliminate the output member from the base class entirely.

> I'm separately trying to figure out the neatest way of letting the base Pipeline (and add_rx_item, shutdown) know about the rx_item_queue

I'm not sure I follow. The `_rx_item_queue` is assigned in `Pipeline.__init__` - does that not achieve it?

amishatishpatel replied:
Agreed on all accounts, but I think it's best if we chat through this one for a few minutes. I'll take you up on your offer of a consult tomorrow morning.

Basically,
* The creation of the item queues requires the `output.name`.
* It doesn't really make sense (to me) to remove the `output` arg (the origin of this comment) and maintain creation of item queues in `Pipeline.__init__`
    * It would require passing through just the `output.name`
    * Which isn't quite desired given idea of one BPipeline to many BOutputs.

amishatishpatel commented:
I think I've addressed the outstanding comments, namely:
* removing `output` arg from `Pipeline.__init__` (and replacing with a Pipeline `name` arg), and
* abstracting `n_{rx, tx}_items` constants into `xbgpu/__init__.py`
    * I'd chosen `xbgpu/__init__.py` instead of the top of `engine.py` because I'd seen `DIG_RMS_DBFS` defaults also placed in `fgpu/__init__` and used in `engine.py` - feel free to tell me to move it.

I want to make some headway on `test_engine.py` before requesting another round of reviewing.

james-smith-za reviewed (commented):
One RFQ. Otherwise I am actually quite happy with this.

I can't figure out where the 1-A problem is but I have a suspicion. I'll do a couple more tests then post an update on JIRA later.

amishatishpatel commented on test/xbgpu/conftest.py:
```diff
@@ -27,17 +29,35 @@ def n_src_streams() -> int:  # noqa: D401
+            for queue, endpoint in zip(stream_queues, endpoints)
+            if IPv4Address(endpoint[0]) in mock_send_stream_network
+        )
+        return spead2.send.asyncio.InprocStream(thread_pool, stream_queues, config)
```
Argh, this has come back to bite me. There are similarities, but they are not identical. I'm tracing through the changes/blame to see what came from where. `test/fgpu/` still references the old/singular **mock_recv_stream** and not *mock_recv_streams*.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -236,116 +271,157 @@ def _create_heaps(
         device_results
             Array of all GPU-generated data of shape
-            - (n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
+            - (len(corrprod_outputs), n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
```
This shape is confusing. Each product can have a different n_total_accumulations, so the actual data doesn't fit into a neat rectangular array. I'd suggest you rather return a list of arrays (one for each corrprod_output).

amishatishpatel replied:
This one grew some tentacles and allowed me to remove the `n_accumulations_completed` list that gets returned alongside this. I've just finished neatening up how it's used in the data verification in the unit test itself, so it might require a bit more engagement during the final review.

Apologies in advance and thank you for this suggestion.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -236,116 +271,157 @@ def _create_heaps(
-        feng_stream = self._make_feng(mock_recv_stream, max_packet_size, max_heaps)
+        feng_stream = self._make_feng(mock_recv_streams, max_packet_size, max_heaps)
+
+        acc_index_counts: list[dict[int, int]] = []
```
Would be helpful to have a comment indicating what this structure represents.

bmerry replied:
I think a comment here would still be helpful, even though it's now a list of sets.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -236,116 +271,157 @@ def _create_heaps(
+        # constructed by the caller.
+        for i, corrprod_output in enumerate(corrprod_outputs):
+            for batch_index in batch_indices:
+                adjusted_index = batch_index // corrprod_output.heap_accumulation_threshold
```
"adjusted" doesn't really describe what adjustment is being made. It looks like it is an accumulation index, so maybe name it `acc_index`.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -236,116 +271,157 @@ def _create_heaps(
-        heap = await recv_stream.get()
-        items = ig_recv.update(heap)
-        assert len(list(items.values())) == 0, "This heap contains item values not just the expected descriptors."
+        for q in mock_recv_streams:
+            q.stop()
```
It's more common in katgpucbf tests to use `queue` for a variable holding an InprocQueue.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -236,116 +271,157 @@ def _create_heaps(
+            for batch_index in batch_indices:
+                adjusted_index = batch_index // corrprod_output.heap_accumulation_threshold
+                acc_index_counts[i][adjusted_index] += 1
+            # `adjusted_index` should be the last accumulation index for this corrprod
```
That comment isn't helping me understand what the code below it is doing. It seems like if the last accumulation is incomplete you remove it from the count array? Why?

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -452,15 +527,34 @@ async def test_xengine_end_to_end(
+            *[corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs]
+        )
+        advance = lowest_common_multiple - (first_batch_index % lowest_common_multiple)
+        first_batch_index += advance
```
This is quite a complicated way to get an index that is a multiple of all the heap sizes. Since there is nothing special about batch 123, you could do something like
```python
first_batch_index = 123 * np.product([corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs])
```

james-smith-za replied:
This was mine actually, it was part of the tweaking that I did to make sure that we were starting the test in a sane place.

I wouldn't suggest further work on this right now - your comment is valid, but rather than delaying this PR further I'd suggest just creating a small technical-debt ticket once we've got it merged.

amishatishpatel replied:
For the record, I've created [NGC-1048](https://skaafrica.atlassian.net/browse/NGC-1048).

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -452,15 +527,34 @@ async def test_xengine_end_to_end(
+            corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs
+        )
+        min_full_accumulations = (
+            lcm(*[corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs])
```
This is the third time you've used `[corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs]`. Things will be more readable if you assign it to a variable once.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -452,15 +527,34 @@ async def test_xengine_end_to_end(
+        range_start = frequency
         range_end = range_start + n_channels_per_substream - 1
-        assert xbengine.sensors["chan-range"].value == f"({range_start},{range_end})"
+        assert xbengine.sensors[f"{corrprod_outputs[0].name}.chan-range"].value == f"({range_start},{range_end})"
```
Check it for every corrprod, not just the first.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -470,7 +564,7 @@ def sensor_observer(sync_sensor: aiokatcp.Sensor, sensor_reading: aiokatcp.Readi
             actual_sensor_updates.append((sensor_reading.value, sensor_reading.status))
 
-        xbengine.sensors["rx.synchronised"].attach(sensor_observer)
+        xbengine.sensors[f"{corrprod_outputs[0].name}.rx.synchronised"].attach(sensor_observer)
```
Check it for every corrprod, not just the first.

amishatishpatel replied:
I didn't expect this to be a quick fix, but I also didn't expect to run into the issue I have. Decided to update this to fill a dictionary (key == sensor name) and the value is a list of tuples.
* Each update is seemingly called once (correctly) - verified with logging statements,
* But the eventual dictionary shows updates from both sensors present in both lists.

Same issue when generating the `expected_sensor_updates` later on. Will keep plugging at it, but just had to share this strangeness.

amishatishpatel replied:
Yeah I figured I hadn't discovered a new branch of quantum entanglement. Thank you for the reminder.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-                mock_recv_stream,
-                recv_stream,
-                heap_accumulation_threshold=heap_accumulation_threshold,
+            batch_start_index = (first_batch_index) * max_heap_acc_threshold - HEAPS_PER_FENGINE_PER_CHUNK
```
Multiplying a batch index by a heap_acc_threshold doesn't make sense.

james-smith-za replied:
Maybe I'm missing something, but ... why not?

james-smith-za replied:
Is `max_heap_acc_threshold` not heaps per batch?
Doesn't change the overall thrust of your argument though, in that you then end up with units of heaps.

So there does seem to be at least some naming inconsistency there.

bmerry replied:
I don't feel confident that the test is doing what it did in `main` (in terms of how it aligns the batches that get sent with the accumulations) since I can't easily work my way through the logic when I don't know what units the variables are in.

If you're confident that the test is doing what it's supposed to do then go ahead and fix it up later (but not much later).

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-                recv_stream,
-                heap_accumulation_threshold=heap_accumulation_threshold,
+            batch_start_index = (first_batch_index) * max_heap_acc_threshold - HEAPS_PER_FENGINE_PER_CHUNK
+            batch_end_index = (first_batch_index + min_full_accumulations) * max_heap_acc_threshold
```
Should probably be
```suggestion
            batch_end_index = first_batch_index + min_full_accumulations * max_heap_acc_threshold
```
Also since `min_full_accumulations * max_heap_acc_threshold` is (currently) just the LCM, this could probably be simplified.

amishatishpatel replied:
As always, fair points. However I'm gona plead the Fifth and cite [James's comment above](https://github.com/ska-sa/katgpucbf/pull/600#discussion_r1296857624) to make this part of a technical-debt-type overhaul.

Your call is still important to us and we thank you for your patience.

james-smith-za replied:
When you create a ticket Amish please just link to these comments. It'll help with the context.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
+        for i, pipeline in enumerate(xpipelines):
+            output_name = pipeline.output.name
+            assert (
+                prom_diff.get_sample_diff("output_x_incomplete_accs_total", {"stream": f"{output_name}"})
```
```suggestion
                prom_diff.get_sample_diff("output_x_incomplete_accs_total", {"stream": output_name})
```

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-        )
-        assert prom_diff.get_sample_diff("output_x_clipped_visibilities_total") == 0
+            assert (
+                prom_diff.get_sample_diff("output_x_heaps_total", {"stream": f"{output_name}"})
```
```suggestion
                prom_diff.get_sample_diff("output_x_heaps_total", {"stream": output_name})
```

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
+                == n_accumulations_completed[i]
+            )
+            # Could manually calculate it here, but it's available inside the send_stream
+            assert prom_diff.get_sample_diff("output_x_bytes_total", {"stream": f"{output_name}"}) == (
```
```suggestion
            assert prom_diff.get_sample_diff("output_x_bytes_total", {"stream": output_name}) == (
```

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -487,64 +582,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
+            assert prom_diff.get_sample_diff("output_x_bytes_total", {"stream": f"{output_name}"}) == (
+                pipeline.send_stream.heap_payload_size_bytes * n_accumulations_completed[i]
+            )
+            assert prom_diff.get_sample_diff("output_x_visibilities_total", {"stream": f"{output_name}"}) == (
```
```suggestion
            assert prom_diff.get_sample_diff("output_x_visibilities_total", {"stream": output_name}) == (
```

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -553,9 +659,9 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
         # will either be all complete or incomplete.
         if missing_antenna is not None:
-            expected_sensor_updates += [(False, aiokatcp.Sensor.Status.ERROR)] * n_full_accumulations
+            expected_sensor_updates += [(False, aiokatcp.Sensor.Status.ERROR)] * (max(n_accumulations_completed) - 1)
```
At present, actual_sensor_updates is collecting from output 0, so this should probably use `n_accumulations_completed[0]` rather than the maximum.

amishatishpatel commented on test/xbgpu/test_engine.py:
```diff
@@ -590,30 +697,34 @@ async def test_saturation(
 
         def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-            timestamp = timestamp_step * batch_index
+            timestamp = batch_index * timestamp_step
```
Indeed indeed, mostly because it bugged me that the `heap_factory` in `test_xengine_end_to_end` (~Line 570) had this timestamp calculation as
> `timestamp = batch_index * timestamp_step`

And this was the other way around (for no apparent reason?).

bmerry replied:
That's a good reason to swap it around then.

james-smith-za commented on test/xbgpu/test_engine.py:
```diff
@@ -336,25 +336,36 @@ async def _send_data(
         feng_stream = self._make_feng(mock_recv_streams, max_packet_size, max_heaps)
 
-        acc_index_counts: list[dict[int, int]] = []
+        # NOTE: The data generation creates an extra heap or several for the
+        # XPipeline with the smallest `heap_accumulation_threshold`. This results
+        # in an incomplete final accumulation for that XPipeline. Incomplete
+        # accumulations are purposefully dictated by `batch_indices` constructed
+        # by the caller.
+        # For these reasons, `acc_index_heap_counts` tracks the heap count for
+        # each accumulation index (for each corrprod output). If a heap count
+        # for a corrprod's final accumulation is not equal to its
+        # `heap_accumulation_threshold`, the accumulation index is discarded
+        # from the list (used later on). If this function attempted to receive
+        # data for an XPipeline's final, incomplete accumulation, it would
+        # result in a _spead2.Stopped error as that accumulation would never
+        # have been processed/transmitted by the XPipeline.
+        # - i.e. the test would attempt to get one more heap than the stream
+        #   contained.
+        acc_index_heap_counts: list[dict[int, int]] = []
```
Is the whole exercise of determining the LCM of the two heap accumulation thresholds not to ensure that this doesn't happen? Or did I miss something?

amishatishpatel replied:
I'd added in this logic to resolve the `spead2 ring buffer has been stopped` error, before we delved further into the 1A debug. I can look into it further now if you want (to see if it's still necessary), but I'm leaning towards making this part of the eventual technical-debt ticket for test cleanup.

james-smith-za replied:
Yeah IMO just leave it for now. It currently works. We will probably have to make several tickets but that's OK.

james-smith-za commented on test/xbgpu/test_engine.py:
```diff
@@ -663,16 +668,21 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
 
-        expected_sensor_updates: list[tuple[bool, aiokatcp.Sensor.Status]] = []
+        # NOTE: Much like initialising `actual_sensor_updates`, a similar
+        # approach must be taken when geenrating `expected_sensor_updates`.
```
typo here I think ("generating")

amishatishpatel commented on test/xbgpu/test_engine.py:
```diff
@@ -307,44 +316,54 @@ async def _send_data(
+                    f"actual: {ig_recv['frequency'].value}."
+                )
+
+                device_results[i][j] = ig_recv["xeng_raw"].value
```
Just btw, `device_results` itself is not a numpy array anymore (as of a few mins ago).

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -602,7 +613,8 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
             incomplete_accums_counter = 0
             base_batch_index = batch_start_index
-            for j in range(n_accumulations_completed[i]):
+            # for j in range(n_accumulations_completed[i]):
```
Remove code rather than commenting it out.

amishatishpatel replied:
Argh, my apologies, thought I cleaned up properly.

bmerry commented on test/xbgpu/test_engine.py:
```diff
@@ -327,58 +327,69 @@ async def _send_data(
+        device_results = [
+            np.zeros(
+                shape=(
+                    len(acc_indices.keys()),  # n_accumulations for this XPipeline
```
I was quite happy with the explicit `.keys` when iterating, since the fact that iterating a dictionary iterates the keys can be a bit non-intuitive. But `len` isn't the same thing as iteration - it just tells you how big the dictionary is (or as you've written it, how big the set of keys in the dictionary is).

Anyway, as mentioned it's just bike-shedding; feel free to leave as is if it makes more sense to you.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +127,417 @@ def reset(self, timestamp: int = 0) -> None:
+    async def gpu_proc_loop(self) -> None:
+        """Perform all GPU processing of received data in a continuous loop.
+
+        This function should do the following:
```
Need a blank line here to make the list render correctly.

bmerry replied:
```suggestion
        This method does the following:
```

I assume you've written "should do" because this is the abstract declaration, but the same docstring gets inherited by the implementation, and it's confusing when consulting the XPipeline documentation. Same applies to the other abstract methods in this class.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +127,417 @@ def reset(self, timestamp: int = 0) -> None:
+    async def sender_loop(self) -> None:
+        """Send heaps to the network in a continuous loop.
+
+        This function should do the following:
```
Need a blank line here to make the list render properly.

bmerry commented on src/katgpucbf/xbgpu/engine.py:
```diff
@@ -117,6 +127,417 @@ def reset(self, timestamp: int = 0) -> None:
+    @abstractmethod
+    def capture_enable(self, enable: bool = True) -> None:
+        """Enable/Disable the transmission of this data product's stream."""
+        raise NotImplementedError
```
Since you've used `#pragma: nocover` a few lines further up, you probably want it here too.

bmerry reviewed (changes requested):
No more code change comments - just some requests to fix up documentation.

amishatishpatel commented:
@bmerry Jenkins is happy with the updates and I can verify the problematic indentation has been fixed (screenshot below). I'm going to hold off on pinging you officially for a review because I want to make sure there are/aren't any code updates required (here) for the `?capture-{start, stop}` change in katsdpcontroller. 

![image](https://github.com/ska-sa/katgpucbf/assets/25218138/18c457f7-8b87-42f4-89bd-c8266780a0c7)
